components/modules/mqtt.c

```diff
@@ -0,0 +1,717 @@
+
+	int n = 6;
+	if( lua_isfunction( L, n ) )
+	{
+		lua_pushvalue( L, n );
+		lua_setfield(L, 1, "_publish_ok");  // set _G["_cb_connect_nok"] = fn()
+		n++;
+	}
+
+	lua_settop(L, top );
+	NODE_DBG("MQTT publish client %p, topic %s, %d bytes\n", client, topic, strlen(data));
+	esp_mqtt_client_publish(client, topic, data, strlen(data), qos, retain);
+	return 0;
+}
+
+// Lua: mqtt:subscribe(topic, qos[, function(client)]) OR mqtt:subscribe(table[, function(client)])
+static int mqtt_subscribe( lua_State* L )
+{
+	esp_mqtt_client_handle_t client = get_client( L );
+
+	int top = lua_gettop(L);
+
+	const char * topic = luaL_checkstring( L, 2 );
+	int qos            = luaL_checkint( L, 3 );
+
+	if( lua_isfunction( L, 4 ) )
+	{
+		lua_pushvalue( L, 4 );
+		lua_setfield(L, 1, "_subscribe_ok");  // set _G["_cb_connect_nok"] = fn()
+	}
+
+	lua_settop(L, top );
+	NODE_DBG("MQTT subscribe client %p, topic %s\n", client, topic);
+	esp_mqtt_client_subscribe(client, topic, qos);
+	return 0;
+}
+
+// Lua: mqtt:unsubscribe(topic[, function(client)]) OR mqtt:unsubscribe(table[, function(client)])
+static int mqtt_unsubscribe( lua_State* L )
+{
+	esp_mqtt_client_handle_t client = get_client( L );
+
+	int top = lua_gettop(L);
+
+	const char * topic = luaL_checkstring( L, 2 );
+	int n = 3;
+	if( lua_isfunction( L, n ) )
+	{
+		lua_pushvalue( L, n );
+		lua_setfield(L, 1, "_unsubscribe_ok");  // set _G["_cb_connect_nok"] = fn()
+		n++;
+	}
+
+	lua_settop(L, top );
+	NODE_DBG("MQTT unsubscribe client %p, topic %s\n", client, topic);
+	esp_mqtt_client_unsubscribe(client, topic);
+	return 0;
+
+	return 0;
+}
+
+static int mqtt_delete( lua_State* L )
+{
+	esp_mqtt_client_config_t * settings = get_settings( L );
+	if( settings != NULL )
+		free( settings );
+
+	esp_mqtt_client_handle_t client = get_client( L );
+	if( client != NULL )
+	{
+		NODE_DBG("stopping MQTT client %p\n", client);
+		esp_mqtt_client_destroy( client );
+		free( client );
+	}
+	return 0;
+}
+
+// Lua: mqtt.Client(clientid, keepalive[, username, password, cleansession])
+static int mqtt_new( lua_State* L )
+{
+	const char * clientid = NULL;
+	clientid = luaL_checkstring( L, 1 );
+	NODE_DBG("MQTT client id %s\n", clientid);
+
+	esp_mqtt_client_config_t * mqtt_cfg = (esp_mqtt_client_config_t *) malloc(sizeof(esp_mqtt_client_config_t));
+	memset(mqtt_cfg, 0, sizeof(esp_mqtt_client_config_t));
+
+	mqtt_cfg->event_handle = mqtt_event_handler;
+
+	strncpy(mqtt_cfg->client_id, clientid, MQTT_MAX_CLIENT_LEN);
+	mqtt_cfg->keepalive = luaL_checkinteger( L, 2 );
+
+	int n = 2;
+	if( lua_isstring(L, 3) )
+	{
+		strncpy( mqtt_cfg->username, luaL_checkstring( L, 3 ), MQTT_MAX_USERNAME_LEN);
+		n++;
+	}
+
+	if( lua_isstring(L, 4) )
+	{
+		strncpy(mqtt_cfg->password, luaL_checkstring( L, 4 ), MQTT_MAX_PASSWORD_LEN);
+		n++;
+	}
+
+	if( lua_isnumber(L, 5) )
+	{
+		mqtt_cfg->disable_clean_session = (luaL_checknumber( L, 5 ) == 0);
+		n++;
+	}
+	lua_pop( L, n ); //remove parameters
+
+	lua_newtable( L );
```

Comment:
I suggest to build the client's metatable not programmatically but from a rotable, e.g. like in the [`ws2812 module`](https://github.com/nodemcu/nodemcu-firmware/blob/dev-esp32/components/modules/ws2812.c#L500).

Reply:
@devsaurus sure thing.  I will look into it.